core-js/src/definitions.ts

```diff
@@ -0,0 +1,1459 @@
+   */
+  *types(includeBuiltIns: boolean = false): Generator<NamedType, void, undefined> {
+    if (includeBuiltIns) {
+      yield* this._builtInTypes.values();
+    }
+    yield* this._types.values();
+  }
+
+  /**
+   * All the built-in types for this schema (those that are not displayed when printing the schema).
+   */
+  builtInTypes(): IterableIterator<NamedType> {
+    return this._builtInTypes.values();
+  }
+
+  /**
+   * The type of the provide name in this schema if one is defined or if it is the name of a built-in.
+   */
+  type(name: string): NamedType | undefined {
+    const type = this._types.get(name);
+    return type ? type : this._builtInTypes.get(name);
+  }
+
+  intType(): ScalarType {
+    return this._builtInTypes.get('Int')! as ScalarType;
+  }
+
+  floatType(): ScalarType {
+    return this._builtInTypes.get('Float')! as ScalarType;
+  }
+
+  stringType(): ScalarType {
+    return this._builtInTypes.get('String')! as ScalarType;
+  }
+
+  booleanType(): ScalarType {
+    return this._builtInTypes.get('Boolean')! as ScalarType;
+  }
+
+  idType(): ScalarType {
+    return this._builtInTypes.get('ID')! as ScalarType;
+  }
+
+  addType<T extends NamedType>(type: T): T {
+    if (this.type(type.name)) {
+      throw buildError(`Type ${type} already exists in this schema`);
+    }
+    if (type.parent) {
+      // For convenience, let's not error out on adding an already added type.
+      if (type.parent == this) {
+        return type;
+      }
+      throw buildError(`Cannot add type ${type} to this schema; it is already attached to another schema`);
+    }
+    if (type.isBuiltIn) {
+      if (!this.isConstructed) {
+        this._builtInTypes.set(type.name, type);
+      } else {
+        throw buildError(`Cannot add built-in ${type} to this schema (built-ins can only be added at schema construction time)`);
+      }
+    } else {
+      this._types.set(type.name, type);
+    }
+    SchemaElement.prototype['setParent'].call(type, this);
+    return type;
+  }
+
+  /**
+   * All the directive defined on this schema _excluding_ the built-in directives, unless explicitly requested.
+   */
+  *directives(includeBuiltIns: boolean = false): Generator<DirectiveDefinition, void, undefined> {
+    if (includeBuiltIns) {
+      yield* this._builtInDirectives.values();
+    }
+    yield* this._directives.values();
+  }
+
+  /**
+   * All the built-in directives for this schema (those that are not displayed when printing the schema).
+   */
+  builtInDirectives(): IterableIterator<DirectiveDefinition> {
+    return this._builtInDirectives.values();
+  }
+
+  directive(name: string): DirectiveDefinition | undefined {
+    const directive = this._directives.get(name);
+    return directive ? directive : this._builtInDirectives.get(name);
+  }
+
+  *allSchemaElement(): Generator<SchemaElement<any, any>, void, undefined> {
+    yield this._schemaDefinition;
```

Comment:
You're right, it's not in `graphql-js` code. That said, the schema definition shares most properties with other elements, namely:
- it can have directive applied to.
- it references other stuffs (types).
- it can have a description.
- it's nice to get it out of `allSchemaElements` to avoid having to special case it (think, code that tries to act on applied directives in a generic way).

I could also argue that, from an pure english language point of view, it _is_ an "element" of a schema. 

So I felt making it a `SchemaElement` was overall a good tradeoff by improving consistency at the price of what I felt was a pretty minor divergence with `graphql-js`.

But I hesitated slightly and I'm amenable to a different opinion. But I will note that if we do change that, we'll have to either:
1. split the current `SchemaElement` into 2 abstractions: `SchemaElement` that would just provide the `coordinate` (it's all it does in `graphql-js`) and another one that has the rest of what `SchemaElement` currently has (applied directives, description, ...). And I'm not sure how to name that new thing. Which adds a new abstraction for users to care about.
2. essentially duplicate a good part of the code of `SchemaElement` into `SchemaDefinition` _and_ change signatures of most methods on `SchemaElement` to be on `SchemaElement | SchemaDefinition`. Which overall sounds ugly.

Reply:
All that said, thinking about it more, I think splitting `SchemaElement` in 2 abstractions might be overall better. I do am unsure how to call it though.

Reply:
Yeah, I struggled with the same question with my AST-based attempt. I considered differentiating between `SchemaElement` and a `NamedSchemaElement` subclass at one point. I don't love it, but `graphql-java` uses [a similar model](https://github.com/graphql-java/graphql-java/blob/2cea0afb4d010a92debcc70cf514b9c446cb32e5/src/main/java/graphql/schema/GraphQLNamedSchemaElement.java#L10).

Reply:
For posterity: I switched to `SchemaElement` and `NamedSchemaElement` naming.